pyispyb/core/routes/datacollections.py

```diff
@@ -0,0 +1,89 @@
+import logging
+import os
+
+from fastapi import Depends, HTTPException, Request, Query
+from fastapi.responses import FileResponse
+from pydantic import conint
+from ispyb import models
+
+from ...dependencies import pagination
+from ...app.extensions.database.utils import Paged
+from ... import filters
+from ...app.base import AuthenticatedAPIRouter
+
+from ..modules import datacollections as crud
+from ..schemas import datacollections as schema
+from ..schemas.utils import paginated
+
+
+logger = logging.getLogger(__name__)
+router = AuthenticatedAPIRouter(prefix="/datacollections", tags=["Data Collections"])
+
+
+@router.get("/images/{dataCollectionId}", response_class=FileResponse)
```

Comment:
I think a REST route format would be more something like "/{datacollectionId}/image/{imageId}"
https://restfulapi.net/resource-naming/

Reply:
This is consistent with getting a list of images and then returning a single image (previous route returns the list)

Reply:
I think we should have two separate routes `/datacollection/attachments` and `datacollection/attachment/{id}` to make it clear in the response model definition wether you should expect to get back a single entity or a list.

Reply:
Or rather `/datacollection/{datacollectionId}/attachments` and `datacollection/{datacollectionId}/attachment/{attachmentId}` since we need to specify the data collection id too

Reply:
Ahh right so maybe:
`/datacollections/{dataCollectionId}/images/?param` => Image

This one needs `dataCollectionId` to return the image (from same table, `DataCollection`)

But 
`/datacollections/attachments`
`/datacollections/attachments/{attachmentId}`

doesn't need `datacollectionId` because we have a unique key `attachmentId` (separate table)

Reply:
I think it is important to have the "s" only when we expect a list:

- `/datacollections/attachments` (with "s") -> list of the attachments (*)
- `/datacollections/attachment/{attachmentId}` (without "s") -> only one attachment

(*) - do we not need the `datacollectionId` here to filter them?

Reply:
> I think it is important to have the "s" only when we expect a list:
> 
> * `/datacollections/attachments` (with "s") -> list of the attachments (*)
> * `/datacollections/attachment/{attachmentId}` (without "s") -> only one attachment
> 
> (*) - do we not need the `datacollectionId` here to filter them?

Thats not what the document you linked says (and other rest best practices i've read say)

> 1.2. Collection and Sub-collection Resources
A resource may contain sub-collection resources also.

> For example, sub-collection resource “accounts” of a particular “customer” can be identified using the URN “/customers/{customerId}/accounts” (in a banking domain).

> Similarly, a singleton resource “account” inside the sub-collection resource “accounts” can be identified as follows: “/customers/{customerId}/accounts/{accountId}“.

`dataCollectionId` is a query parameter, because you may want to filter them by other things such as  `dataCollectionGroupId` too

Reply:
Ah yes you're right, I missread it. My bad!

Reply:
In this case, should the endpoints be:
`/attachments`   
`/attachments/{attachmentId}` 
(ie. remove `/datacollections`)?
with the query params `?dataCollectionId={}` or `?dataCollectionGroupId={}`  or whatever  
as `/datacollections/attachments?datacollectionId={id}` does not really follow the api convention naming.

Reply:
I think `/datacollections/attachments` makes sense as it makes it clear that it is data collection attachments that we are getting (could be auto proc attachments or something else). Why do you say it does not really follow the api convention naming?

Reply:
attachments are a sub-collection of datacollection (db: `DataCollectionFileAttachment.dataCollectionId` -> `DataCollection`) hence the tree, but you dont have to filter on dataCollectionId. Without datacollection how do you know what attachments are? (we have AutoProcProgram attachments too, and others...)

Reply:
As far I understand, it should be either:
`datacollections/{dataCollectionId}/attachments` 
or
`/attachments` with some query params 
or `/datacollectionattachments` if `/attachments` is not understandable  

For example, you will not create `/shippings/dewars` if you want to query all dewars for either a given shippingId or a give name or another param, you will rather create `/dewars` with some query params (shippingId or name). Or you will create `/shipping/{id}/dewars` .

Reply:
I dont think i agree, the path indicates the entity hierarchy as well. `/datacollectionattachments` is pretty horrible as far as readability goes

Take a more complex example: autoproc, we have other linked tables: `AutoProcProgram`, `AutoProcProgramAttachments`, `AutoProcProgramMessages`

i would define:
```
/autoprocs/
/autoprocs/{autoprocprogramid}
/autoprocs/messages
/autoprocs/messages/{messageid}
/autoprocs/attachments
/autoprocs/attachments/{attachmentid}
```

you want these resources to be grouped together, even though you may not filter on `{autoprocprogramid}`, for example you might want to show all messages for a `datacollection` as well as for a specific `autoprocprogramid`

Reply:
I agree it should be great to group them together, but I would have chosen  a name different from the name of an existing resource endpoint. For example `/processing`, 
```
/processing/autoprocs
/processing/messages
/processing/attachments
```
otherwise we could have long urls, if we really want to represent the sub collections, like `/autoprocs/programs/attachments`
but I'm okay if you prefer have the same name, it was just a comment.